### ISSUE TITLE
Annotate references with the name of the function in which they occur

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#177]
When finding references, it would be really useful if each reference was annotated with the name of the function in which the reference occurs (in addition to the name of the file, which is already there). 

For example, given this code:

```
void target();

void function1() {
  target();
}

void function2() {
  target();
}
```

if you perform "Find References" on `target()`, I would like an editor to be able to show output similar to this:

![cfviz-refs-eclipse](https://user-images.githubusercontent.com/1751085/34898433-10f6ab8c-f7c1-11e7-8866-e95c4be74d47.png)

Notice how, beside each reference, there is the name of the function in which the reference occurs.

In terms of implementation, this likely needs the same sort of infrastructure as [call hierarchy](https://github.com/clangd/clangd/issues/162) (storing a `SymbolID` inside `Ref`).

**Comment by kadircet** [comment]
Today LSP doesn't have a way of attaching "snippets/details" to references response (it's just a list of locations). this might be implemented as an extension, but on the client side requires extra work to display that detail and most of the clients already display the text at the reference side as a snippet.

So I don't think this is worth as providing an extension today, even though we have all the infra ready. WDYT @HighCommander4 ?

**Comment by kadircet** [comment]
(oops didn't mean to close)

**Comment by HighCommander4** [comment]
In my experience, this is one of the most noticeable remaining feature gaps between clangd and Eclipse CDT. The "code surrounding the reference" snippet helps somewhat, but is only a few lines long and typically needs to be scrolled to bring function name into view.

I'd like to work on an extension that implements this (including an LSP extension), just haven't gotten around to it yet.

**Comment by tom-anders** [comment]
I found this idea quite interesting and wanted to try it out at work next week, so I threw together a quick and dirty prototype here: https://reviews.llvm.org/D137894

@HighCommander4 Are you still planning on working on this?

**Comment by HighCommander4** [comment]
@tom-anders I would love to have this! I haven't had a chance to get to it myself -- if you're interested in working on it, please feel free!

Note that it will require client-side changes too (until a suitable LSP extension is accepted upstream).

**Comment by tom-anders** [comment]
Cool, I'll polish up the patch a bit in the next few days then :)

Regarding client side changes: I'm using neovim's builtin LSP, so I'd be happy to adapt [clangd_extensions.nvim](https://github.com/p00f/clangd_extensions.nvim) for this. Someone else would need to adapt the vscode plugin though.

**Comment by sam-mccall** [comment]
Yeah I can believe this is really useful, refs is a heavily-used feature and benefits from context for quickly scanning the list.

However the VSCode extension API (which their LSP library implements) has no facility to include this information, it's just a list of Locations. So I think this can't be implemented in vscode-clangd without reinventing the several find-refs UIs in VSCode (which I think is probably a bad idea outside prototyping, for maintenance reasons).

So this likely needs to be a feature request to VSCode itself rather than just LSP (I think it's a good one, but that doesn't mean it'll be added). Happy to vote for such a FR!

IMO this shouldn't block adding the extension to clangd: it seems useful and unintrusive enough even without vscode support.
